man/GUI-tools/csettool.man
.so ../macros
.TH csettool GUI-tools %G% "BitMover, Inc." BitKeeper
.\"    =============== BitKeeper Graphical ChangeSet Browser ===============
.SH NAME
bk csettool \- graphical changeset browser
.SH SYNOPSIS
.if t \{\
.B bk csettool 
.BI -r revs
\}
.if n \{\ 
bk csettool -r<revs>
\}
.SH DESCRIPTION
.LP
The csettool command is used to view the detailed information about the
specified changeset[s].  The tool displays the list of changes in each
changeset, the ChangeSet history, and (optionally) the differences found
in each file contained in each changeset.
.LP
To see the changes for a particular file, click on the file name
in the top left window and you will see:
.TP
.B \-
the number of diffs in the light blue, middle status window
.TP
.B \-
the old version of the file in the lower left window
.TP
.B \-
the new version of the file in the lower right window
.TP
.B \-
the ChangeSet's comments followed by the delta's comments in
the upper right window
.LP
Use the space bar to move between diffs and files.  Each time you
hit the space bar, the next diff is brought into view.  If you are
on the last diff, the tool moves to the next file. The Next (>>)
buttons and Previous (<<) buttons in the upper left corner will also
allow you to browse the files and diffs.
.SH BINDINGS
.TP 16
.B Home
Scroll both files to the top
.tp
.B End
Scroll both files to the bottom
.tp
.B PageUp
Scroll both files one screen up
.tp
.B PageDown
Scroll both files one screen down
.tp
.B
UpArrow
Scroll both files one line up
.tp
.B DownArrow
Scroll both files one line down
.tp
.B LeftArrow
Scroll both files to the left
.tp
.B RightArrow
Scroll both files to the right
.tp
.B Alt-UpArrow
Make the diffs windows one line bigger and
the lists windows one line smaller.
.tp
.B Alt-DownArrow
Make the diffs windows one line smaller and
the lists windows one line bigger.
.tp
.B q
Quit
.tp
.B space
Next diff, if last diff, then next file
.tp
.B n
Next diff, if last diff, then next file
.tp
.B p
Previous diff, if first diff, then previous file
.tp
.B .
Center the current diff on the screen
.tp
.B N
Next file
.tp
.B P
Previous file
.SH "FUTURE DIRECTIONS"
.LP
We plan on having a single screen diff window option where the changes
are all shown color coded in one window.  
.SH "SEE ALSO"
.B bk help cset
.br
.B bk help difftool
.br
.B bk help ranges


RELEASE-NOTES
============================================================================
Release notes for BitKeeper version 6.0.2 (released Oct 29, 2013)

    This is a bugfix release (one feature snuck in).

Bugfixes

  bk diffs
    - Fixed a bug where 'bk diffs -NL' did not list extra files.
    - bk diffs -s was not working because of the switch to -u as
      the default.
    - Function header matching in 'bk diffs -p' now works with C++
      names.

  bk difftool
    - When looking at diffs over different types of files, a binary
      file in the list would cause diffs for the next text file to not
      be displayed.

  bk gone
    - People would try and run 'bk gone foo.c' which is incorrect, the
      correct usage is:
	bk gone `bk log -r+ -nd:ROOTKEY:`
      bk gone now works harder to validate that the arguments passed
      are actually keys.

  Other bugfixes
    - Fix a bug in the BK6->BK5 compat code where some files were not
      converted correctly to the BK5 format.
    - Stale locks on windows were not always being cleaned immediately.
    - Ignore extra files in .bk which were causing problems in Mac OS.

Features

   bk annotate
    - MD5 keys are now supported in the annotation options by running
      either 'bk annotate -a5' or 'bk annotate -A5'.

============================================================================
Release notes for BitKeeper version 6.0.1 (released Aug 9, 2013)

    This is a bugfix release (a few features snuck in).

Bugfixes

  bk clone
    - 'bk clone -@' wasn't populating comps when the remote repository
      was a subset of the local baseline.
    - 'bk clone -r<rev> [--upgrade|--downgrade] repo localcopy' could
      fail in some cases, and has been fixed.
    - 'bk clone -r<rev> bk://host/repo localcopy' will now fail if the
      BK client version is bk-6.0 and the bkd is bk-6.0.1 or later.
      The solution is to upgrade the client.

  bk commit
    - 'bk commit -f' was still prompting for comments.
    - 'bk commit' had lost some status reporting; that has been
      restored.

  bk collapse
    - 'bk collapse' would leave components marked as pending; this caused
      pull to error with a "KEY MISMATCH" message.
    - 'bk fix/collapse' in a gate was only partially disabled, it would
      collapse in the product but would refuse to collapse in
      components.

  bk diffs
    - 'bk -Ucx diffs -N' would only print diffs for one new file;
      fixed to print diffs for all new files

  bk difftool
    - The file listing box on difftool resizes itself to be as large
      as the longest pathname shown (it used to truncate).
    - 'bk difftool -S -Lpath/to/component' is now supported.
      (-S == scan only the current component)
    - The 'next' button in bk difftool didn't come up highlighted when
      there was more than one file.
    - Better error reporting with -L when a single GCA doesn't exist.

  bk pull
    - A pull from a remote repository served by a bk version 5.x could fail
      if a component clone was part of the operation.  It works now.

  bk push
    - A push from a local repository served by a bk version 5.x to a remote
      repository that uses new storage format of bk-6 could fail if a component
      clone was part of the operation.  It works now.

  bk revtool
    - A particular sequence of clicks and launching difftool could
      make revtool unresponsive.

    - When launching revtool with a line number (bk revtoo +lno),
      the comments for the revision that added that line were not being shown.

  bk sfiles
    - In rare cases, a pending component would not be seen as pending
      ('bk -pA' did not list them but it does now).

  Misc
    - In rare cases, permissions in the .bk/SCCS directory could be set
      wrong and cause bk to fail.  The known case has been fixed and
      the problem is now repaired automatically.

    - In another permissions failure, when a user didn't have write
      permissions on the repository, commit would fail with an
      unhelpful error message.  The problem is now reported more
      clearly.

  Graphical tools
    - When resizing graphical tools quickly, the resize labels could
      get out of sync; they update more frequently now.

Features
    - bk difftool now supports -w/-b to ignore white space changes
      like 'bk diffs'.  There is toolbar control of this feature as
      well.

    - BitKeeper versions prior to 6.0 may now clone repositories in the new
      format; they will automatically be downgraded to a compatible format.

============================================================================
Release notes for BitKeeper version 6.0 (released Jul 05, 2013)

    The primary focus of the 6.0 effort is performance improvements.

New repository format for performance
    BK6 has a new file format designed for performance.

    The new repository format cannot be read by bk-5.x releases
    of BitKeeper, but clone can convert a new repository to a bk-5.x
    readable repository.

Compatibility
    New file formats are a cause for concern but we've done the extra
    work so you may try it without forcing everyone to upgrade. BK6 can
    pull/push the new format from/to bk5's format.

    To enable the the new, faster format:
  
	bk clone --upgrade bk5-repo bk6-repo
  
    To downgrade to be compatible with bk5 (or bk4 if it is a non-nested
    repo):

	bk clone --downgrade bk6-repo bk5-repo


src/gui/csettool.tcl
# diffrtool - view differences between repositories
# Copyright (c) 1999 by Larry McVoy; All rights reserved

# If even partially visible, return 1
proc nextFile {} \
{
	global	fileCount lastFile

	if {$lastFile == $fileCount} { return }
	incr lastFile
	dotFile
}

proc redoFile {} \
{
	# Preserve the current view, re-do all the magic, then restore
	# the view
	set view [diffView]
	dotFile
	diffView $view
	
}

proc prevFile {} \
{
	global	lastFile

	if {$lastFile == 1} { return 0 }
	incr lastFile -1
	dotFile
	return 1
}

# XXX: Some functionality that Larry never implemented?
proc nextCset {} \
{
}

proc prevCset {} \
{
}

proc file_history {} \
{
	global	lastFile Files file_stop RealFiles

	set line $Files($lastFile)
	set line "$line.0"
	set file $RealFiles($lastFile)
	regexp "^  $file_stop" "$file" -> file rev
	catch {exec bk -R revtool -l$rev "$file" &}
}

# Takes a line number as an arg when creating continuations for the file menu
proc dotFile {{line {}}} \
{
	global	lastFile fileCount Files file_stop
	global	RealFiles file finfo currentCset dev_null
	global	gc

	set finfo(lt) ""
	set finfo(rt) ""
	if {$line != ""} { set lastFile $line }
	if {$lastFile == 1} {
		.menu.prevFile configure -state disabled
	} else {
		.menu.prevFile configure -state normal
	}
	if {$lastFile == $fileCount} {
		.menu.nextFile configure -state disabled
	} else {
		.menu.nextFile configure -state normal
	}
	set line $Files($lastFile)
	set line "$line.0"
	.l.filelist.t see $line
	.l.filelist.t tag remove select 1.0 end
	.l.filelist.t tag add select $line "$line lineend + 1 char"
	set file $RealFiles($lastFile)

	clearInfo "Working..."

	# busy is put after we change the selection. This is because busy
	# causes a screen update and we want the selection set quickly to make
	# the user think we're responsive.
	busy 1
	regexp "^  $file_stop" "$file" -> file rev
	set    dspec ":PARENT:\\n"
	append dspec ":T|PARENT: :Dd|PARENT::DM|PARENT::Dy|PARENT:\\n"
	append dspec ":T: :Dd::DM::Dy:"
	set p [open "| bk prs -hr$rev {-nd$dspec} \"$file\""]
	gets $p parent
	gets $p finfo(lt)
	gets $p finfo(rt)
	catch {close $p}
	if {$parent == ""} { set parent "1.0" }
	set finfo(l) "$file@$parent"
	set finfo(r) "$file@$rev"
	set l [tmpfile csettool]
	set r [tmpfile csettool]
	if {$::showAnnotations} {
		set annotate "$gc(cset.annotation)"
		if {[string index $annotate 0] != "-"} {
			set annotate "-$annotate"
		}
		if {[string first "a" $annotate] == -1} {
			append annotate "a"
		}
		if {$annotate == "-a"} {set annotate "-aum"}
	} else {
		set annotate ""
	}

	set buf ""
	set line [lindex [split $line "."] 0]
	while {[regexp {^ChangeSet (.*)$} $buf dummy crev] == 0} {
		incr line -1
		set buf [.l.filelist.t get "$line.0" "$line.0 lineend"]
	}
	set currentCset [lindex [split $buf] 1]
	.l.sccslog.t configure -state normal
	.l.sccslog.t delete 1.0 end

	set dspec \
	    "-d:GFILE: :I: :D: :T: :P:\$if(:HT:){@:HT:}\\n\$each(:C:){  (:C:)\\n}"
	set prs [open "| bk prs {$dspec} -hr$crev ChangeSet" r]
	set first 1
	while { [gets $prs buf] >= 0 } {
		if {$first == 1} {
			set first 0
			.l.sccslog.t insert end "$buf\n" cset
		} else {
			.l.sccslog.t insert end "$buf\n"
		}
	}
	catch { close $prs }

	set prs [open "| bk prs -bhC$rev {$dspec} \"$file\"" r]
	set save ""
	while { [gets $prs buf] >= 0 } {
		if {$buf == "  "} { continue }
		if {[regexp {^  } $buf]} {
			if {$save != ""} {
				.l.sccslog.t insert end "$save\n" file_tag
				set save ""
			}
			.l.sccslog.t insert end "$buf\n"
		} else {
			# Save it and print it later iff we have comments
			set save $buf
		}
	}
	catch { close $prs }
	while {[.l.sccslog.t get "end - 2 char" end] == "\n\n"} {
		.l.sccslog.t delete "end - 1 char" end
	}
	.l.sccslog.t configure -state disabled
	.l.sccslog.t see end
	.l.sccslog.t xview moveto 0
	update idletasks

	if {$annotate == ""} {
		catch { exec bk get -qkpr$parent "$file" > $l}
		catch { exec bk get -qkpr$rev "$file" > $r}
	} else {
		catch { exec bk get -qkpr$parent $annotate "$file" > $l}
		catch { exec bk get -qkpr$rev $annotate "$file" > $r}
	}
	if {[isChangeSetFile $file] && [isComponent $file]} {
		catch {
			exec bk changes -S -vr$rev [file dirname $file] > $r
		} result
		readFiles $dev_null $r
	} else {
		displayInfo $file $file $parent $rev 
		readFiles $l $r
	}

	busy 0
}

proc savePatch {f s} \
{
	global out
	puts -nonewline $out $s
}

proc exportCset {} \
{
	global bgExec out currentCset

	if {![info exists currentCset] || $currentCset eq ""} {
		displayMessage "You are not in a ChangeSet that can be exported"
		return
	}
	set f [tk_getSaveFile -title "Select file for patch export..."]
	if {$f eq ""} return
	busy 1
	set out [open $f w+]
	set r [bgExec -output savePatch bk export -tpatch -r$currentCset]
	if {$r} {
		displayMessage "Export Failed: $bgExec(stderr)"
	} elseif {[catch {close $out} e]} {
		displayMessage "Export failed: $e"
	} else {
		# success!
		displayMessage "Export completed successfully\n$f"
	}
	busy 0
}

proc getFiles {revs {file_rev {}}} \
{
	global	fileCount lastFile Files line2File
	global  RealFiles fmenu file_old_new bk_fs
	global	dashs

	busy 1

	# Only search for the last part of the file. This might fail when 
	# there are multiple file.c@rev in the tree. However, I am trying 
	# to solve the case where csettool is called like 
	# -f~user/some_long_path/src/file.c. The preceding would never 
	# match any of the items in the file list
	set file_rev [file tail $file_rev]

	# Initialize these variables so that files with no differences don't
	# cause failures
        set Diffs(0) 1.0
        set DiffsEnd(0) 1.0

	.l.filelist.t configure -state normal
	.l.filelist.t delete 1.0 end
	set fileCount 0
	set line 0
	set found ""
	set match ""
	set S ""
	if {$dashs} { set S "-S" }
	if {$revs == "-"} {
		set r [open "| bk changes $S -faevnd:GFILE:|\$if(:DT:!=D)\{TAGS:\$each(:TAG:)\{(:TAG:),\}\}\$if(:DT:=D)\{:DPN:\}|:I: --no-meta -" r]
	} else {
		set r [open "| bk changes $S -fvnd:GFILE:|:DPN:|:REV: -r$revs" r]
	}
	set csets 0
	set tags 0
	set t [clock clicks -milliseconds]
	while {[gets $r buf] > 0} {
		regexp  $file_old_new $buf dummy name oname rev
		if {[string match "1.0" $rev]} continue
		if {$name == "ChangeSet"} {
			if {[string match TAGS:* $oname]} {
				incr tags
				continue
		    	}
			.diffs.status.middle \
			    configure -text "Getting cset $csets"
			set now [clock clicks -milliseconds]
			if {$now - $t > 200} {
				update
				set t $now
			}
			.l.filelist.t insert end "ChangeSet $rev\n" cset
			incr csets
			incr line
			continue
		}
		incr line
		incr fileCount
		set line2File($line) $fileCount
		set Files($fileCount) $line

		set RealFiles($fileCount) "  $name@$rev"
		set buf "$oname@$rev"
		if {[string first $file_rev $buf] >= 0} {
			set found $fileCount
		}
		.l.filelist.t insert end "  $buf\n"
		$fmenu(widget) add command -label "$buf" \
		    -command  "dotFile $fileCount"
	}
	catch { close $r }
	if {$revs ne "-" && !$fileCount} {
		displayMessage "This ChangeSet is a merge ChangeSet and does\
		    not contain any files."
		exit
	}
	if {($tags > 0) && ($csets == 0)} {
		global	env

		set x [expr [winfo rootx .diffs.status.middle] - 50]
		set y [expr [winfo rooty .diffs.status.middle] + 10]
		set env(BK_MSG_GEOM) "+$x+$y"
		exec bk prompt -i -o -G "No changesets found, only tags"
	}
	if {$fileCount == 0} {
		exit
	}
	.l.filelist.t configure -state disabled
	set lastFile 1
	wm title . "Cset Tool - viewing $csets changesets"
	if {$found != ""} {
		dotFile $found
	} else {
		dotFile
	}
	busy 0
}

# --------------- Window stuff ------------------

# the purpose is to clear out all the widgets; typically right before
# filling them back up again.
proc clearInfo {{message ""}} \
{
	.diffs.status.middle configure -text $message
	.diffs.status.l configure -text ""
	.diffs.status.r configure -text ""
	.diffs.left configure -state normal
	.diffs.right configure -state normal
	.l.sccslog.t configure -state normal
	.diffs.left delete 1.0 end
	.diffs.right delete 1.0 end
	.l.sccslog.t delete 1.0 end
	.diffs.left configure -state disabled
	.diffs.right configure -state disabled
	.l.sccslog.t configure -state disabled
}

proc busy {busy} \
{
	set oldCursor [. cget -cursor]
	if {$busy == 1} {
		. configure -cursor watch
		.l.filelist.t configure -cursor watch
		.l.sccslog.t configure -cursor watch
		.diffs.left configure -cursor watch
		.diffs.right configure -cursor watch
	} else {
		. configure -cursor left_ptr
		.l.filelist.t configure -cursor left_ptr
		.l.sccslog.t configure -cursor left_ptr
		.diffs.left configure -cursor left_ptr
		.diffs.right configure -cursor left_ptr
	}
	# only call update if the cursor changes; this will cut down
	# a little bit on the flashing that happens at startup. It doesn't
	# eliminate the problem, but it helps. 
	if {![string match $oldCursor [. cget -cursor]]} {
		update
	}
}

proc pixSelect {x y} \
{
	global	lastFile line2File file

	set line [.l.filelist.t index "@$x,$y"]
	set x [.l.filelist.t get "$line linestart" "$line linestart +2 chars"]
	if {$x != "  "} { return }
	set line [lindex [split $line "."] 0]
	# if we aren't changing which line we're on there's no point in
	# calling dotFile since it is a time consuming process
	if {$line2File($line) == $lastFile} {return}
	set lastFile $line2File($line)
	dotFile
}

proc adjustHeight {diff list} \
{
	global	gc 

	incr gc(cset.listHeight) $list
	.l.filelist.t configure -height $gc(cset.listHeight)
	.l.sccslog.t configure -height $gc(cset.listHeight)
	incr gc(cset.diffHeight) $diff
	.diffs.left configure -height $gc(cset.diffHeight)
	.diffs.right configure -height $gc(cset.diffHeight)
}

proc gotoProductCset {} \
{
	global	currentCset dashs

	if {[catch {exec bk r2c -r$currentCset ChangeSet} productCset]} {
	    popupMessage -E "Could not find Product ChangeSet revision."
	    return
	}

	pack forget .menu.product
	cd2product
	set dashs 0
	getFiles $productCset ""
}

proc widgets {} \
{
	global	scroll gc wish d search fmenu app env

	getConfig "cset"

	set gc(bw) 1
	set gc(mbwid) 8
	if {$gc(windows)} {
		set gc(py) 0; set gc(px) 1
	} elseif {$gc(aqua)} {
		set gc(py) 3; set gc(px) 10
		set gc(mbwid) 4
	} else {
		set gc(py) 1; set gc(px) 4
	}

	ttk::frame .l
	ttk::frame .l.filelist
	    text .l.filelist.t -height $gc(cset.listHeight) -width 40 \
		-state disabled -wrap none -font $gc(cset.fixedFont) \
		-setgrid true \
		-xscrollcommand { .l.filelist.xscroll set } \
		-yscrollcommand { .l.filelist.yscroll set } \
		-background $gc(cset.listBG) -foreground $gc(cset.textFG)
	    ttk::scrollbar .l.filelist.xscroll -orient horizontal \
		-command ".l.filelist.t xview"
	    ttk::scrollbar .l.filelist.yscroll -orient vertical \
		-command ".l.filelist.t yview"
	    grid .l.filelist.t -row 0 -column 0 -sticky news
	    grid .l.filelist.yscroll -row 0 -column 1 -sticky nse 
	    grid .l.filelist.xscroll -row 1 -column 0 -sticky ew
	    grid rowconfigure    .l.filelist .l.filelist.t -weight 1
	    grid columnconfigure .l.filelist .l.filelist.t -weight 1

	ttk::frame .l.sccslog
	    text .l.sccslog.t -height $gc(cset.listHeight) -width 80 \
		-state disabled -wrap none -font $gc(cset.fixedFont) \
		-setgrid true \
		-xscrollcommand { .l.sccslog.xscroll set } \
		-yscrollcommand { .l.sccslog.yscroll set } \
		-background $gc(cset.listBG) -foreground $gc(cset.textFG)
	    ttk::scrollbar .l.sccslog.xscroll -orient horizontal \
		-command ".l.sccslog.t xview"
	    ttk::scrollbar .l.sccslog.yscroll -orient vertical \
		-command ".l.sccslog.t yview"
	    grid .l.sccslog.t -row 0 -column 0 -sticky news
	    grid .l.sccslog.yscroll -row 0 -column 1 -sticky ns
	    grid .l.sccslog.xscroll -row 1 -column 0 -sticky ew
	    grid rowconfigure    .l.sccslog .l.sccslog.t -weight 1
	    grid columnconfigure .l.sccslog .l.sccslog.t -weight 1

	    createDiffWidgets .diffs

	set prevImage [image create photo \
			   -file $env(BK_BIN)/gui/images/previous.gif]
	set nextImage [image create photo \
			   -file $env(BK_BIN)/gui/images/next.gif]

	ttk::frame .menu
	    ttk::button .menu.prevCset -text "<< Cset" -command prevCset
	    ttk::button .menu.nextCset -text ">> Cset" -command nextCset
	    ttk::button .menu.prevFile -image $prevImage -command prevFile
	    ttk::menubutton .menu.fmb -text "File" -menu .menu.fmb.menu
		set fmenu(widget) [menu .menu.fmb.menu]
		if {$gc(aqua)} {$fmenu(widget) configure -tearoff 0}
		$fmenu(widget) add checkbutton \
		    -label "Show Annotations" \
		    -onvalue 1 \
		    -offvalue 0 \
		    -variable showAnnotations \
		    -command redoFile
		$fmenu(widget) add separator
	    ttk::button .menu.nextFile -image $nextImage -command nextFile
	    ttk::button .menu.prev -image $prevImage -state disabled -command {
		searchreset
		prev
	    }
	    ttk::button .menu.next -image $nextImage -state disabled -command {
		searchreset
		next
	    }
	    ttk::menubutton .menu.mb -text "History" -menu .menu.mb.menu
		set m [menu .menu.mb.menu]
		if {$gc(aqua)} {$m configure -tearoff 0}
		$m add command -label "Changeset History" \
		    -command "exec bk revtool &"
		$m add command -label "File History" \
		    -command file_history
	    ttk::button .menu.quit -text "Quit" -command exit 
	    ttk::button .menu.help -text "Help" -command {
		exec bk helptool csettool &
	    }
	    if {[inComponent] && ![inRESYNC]} {
		ttk::button .menu.product -text "View Product" -command {
		    gotoProductCset
		}
	    }
	    ttk::button .menu.dot -text "Current diff" -command dot

	    pack .menu.quit -side left -padx 1
	    pack .menu.help -side left -padx 1
	    if {[winfo exists .menu.product]} {
		    pack .menu.product -side left -padx 1
	    }
	    pack .menu.mb -side left -padx 1
	    pack .menu.prevFile -side left -padx 1
	    pack .menu.fmb -side left -padx 1
	    pack .menu.nextFile -side left -padx 1
	    pack .menu.prev -side left -padx 1
	    pack .menu.dot -side left -padx 1
	    pack .menu.next -side left -padx 1
	    # Add the search widgets to the menu bar
	    search_widgets .menu .diffs.right

	    # We put status info in the diff status window that is larger
	    # than that expeced by difflib; so, make the label a bit wider 
	    # to keep the display from jiggling
	    .diffs.status.middle configure -width 25
	
	# smaller than this doesn't look good.
	#wm minsize . $x 400

	grid .menu -row 0 -column 0 -sticky ew -pady 2
	grid .l -row 1 -column 0 -sticky nsew
	grid .l.sccslog -row 0 -column 1 -sticky nsew
	grid .l.filelist -row 0 -column 0 -sticky nsew
	grid .diffs -row 2 -column 0 -sticky nsew
	grid rowconfigure .menu 0 -weight 0
	grid rowconfigure .diffs 1 -weight 1
	grid rowconfigure . 0 -weight 0
	grid rowconfigure . 1 -weight 0
	grid rowconfigure . 2 -weight 2

	grid rowconfigure    .l .l.filelist -weight 1
	grid columnconfigure .l .l.filelist -weight 1
	grid rowconfigure    .l .l.sccslog  -weight 1
	grid columnconfigure .l .l.sccslog  -weight 1

	grid columnconfigure . 0 -weight 1
	grid columnconfigure .menu 0 -weight 1
	grid columnconfigure .diffs 0 -weight 1

	#$search(widget) tag configure search \
	#    -background $gc(cset.searchColor) -font $gc(cset.fixedBoldFont)
	keyboard_bindings
	search_keyboard_bindings
	searchreset
	foreach w {.diffs.left .diffs.right} {
		bindtags $w {all Text .}
	}

	computeHeight "diffs"

	.l.filelist.t tag configure select -background $gc(cset.selectColor) \
	    -relief groove -borderwid 1
	.l.filelist.t tag configure cset \
	    -background $gc(cset.listBG) -foreground $gc(cset.textFG)
	.l.sccslog.t tag configure cset -background $gc(cset.selectColor) 
	.l.sccslog.t tag configure file_tag -background $gc(cset.selectColor) 
	. configure -cursor left_ptr
	.l.sccslog.t configure -cursor left_ptr
	.l.filelist.t configure -cursor left_ptr
	.diffs.left configure -cursor left_ptr
	.diffs.right configure -cursor left_ptr
}

# Set up keyboard accelerators.
proc keyboard_bindings {} \
{
	global afterId gc search

	bind all <Control-b> { if {[Page "yview" -1 0] == 1} { break } }
	bind all <Control-f> { if {[Page "yview"  1 0] == 1} { break } }
	bind all <Control-e> { if {[Page "yview"  1 1] == 1} { break } }
	bind all <Control-y> { if {[Page "yview" -1 1] == 1} { break } }
	bind all <Prior> { if {[Page "yview" -1 0] == 1} { break } }
	bind all <Next> { if {[Page "yview" 1 0] == 1} { break } }
	bind all <Up> { if {[Page "yview" -1 1] == 1} { break } }
	bind all <Down> { if {[Page "yview" 1 1] == 1} { break } }
	bind all <Left> { if {[Page "xview" -1 1] == 1} { break } }
	bind all <Right> { if {[Page "xview" 1 1] == 1} { break } }
	bind all <Home> {
		global	lastDiff

		set lastDiff 1
		dot
		.diffs.left yview -pickplace 1.0
		.diffs.right yview -pickplace 1.0
	}
	bind all <End> {
		global	lastDiff diffCount

		set lastDiff $diffCount
		dot
		.diffs.left yview -pickplace end
		.diffs.right yview -pickplace end
	}
	bind all <Alt-Up> { adjustHeight 1 -1 }
	bind all <Alt-Down> { adjustHeight -1 1 }
	bind all <$gc(cset.quit)>	exit
	bind all <space>		next
	bind all <n>			next
	bind all <p>			prev
	bind all <r>			file_history
	bind all <period>		dot
	bind all <Control-n>		nextFile
	bind all <Control-p>		prevFile
	bind all <s>			exportCset

	if {$gc(aqua)} {
		bind all <Command-q> exit
		bind all <Command-w> exit
	}
	# note that the "after" is required for windows. Without
	# it we often never see the double-1 events. 
	bind .l.filelist.t <1> { 
		set afterId \
		    [after idle [list after $gc(cset.doubleclick) \
				     pixSelect %x %y]]
		break
	}
	# the idea is, if we detect a double click we'll cancel the 
	# single click, then make sure we perform the single and double-
	# click actions in order
	bind .l.filelist.t <Double-1> {
		if {[info exists afterId]} {
			after cancel $afterId
		}
		pixSelect %x %y
		file_history
		break
	}
	# In the search window, don't listen to "all" tags.
	bindtags $search(text) { .menu.search TEntry . }
}

proc usage {} \
{
	catch {exec bk help -s csettool} usage
	puts stderr $usage
	exit 1
}


proc main {} \
{
	global argv0 argv argc app showAnnotations gc dashs

	wm title . "Cset Tool"

	# Set 'app' so that the difflib code knows which global config
	# vars to read
	set revs ""
	set argindex 0
	set file_rev ""
	set dashs 0
	set stdin 0

	while {$argindex < $argc} {
		set arg [lindex $argv $argindex]
		switch -regexp -- $arg {
		    "^-f.*" {
			set ftmp [lindex $argv $argindex]
		   	regexp {^[ \t]*-f(.*)} $ftmp dummy file_rev
		    }
		    "^-r.*" {
			set rev [lindex $argv $argindex]
		   	regexp {^[ \t]*-r(.*)} $rev dummy revs
			# make sure we don't get an empty revision
			if {$revs eq ""} { usage }
		    }
		    "^-S$" {
			set dashs 1
		    }
		    "^-$" {
			set stdin 1
		    }

		    default {
			if {$file_rev ne ""
			    || $argindex != [expr {$argc - 1}]
			    || [catch {cd $arg} err]} { usage }
		    }
		}
		incr argindex
	}
	if {(($revs != "") || ($file_rev != "")) && $stdin} {
		wm withdraw .
		displayMessage "Can't use '-' option with any other options"
		exit
	}
	if {$revs == ""} {
		set revs "+"
	}
	#displayMessage "csetttool: revs=($revs) file=($file_rev)"
	bk_init

	if {$dashs} {
		if {[cd2root [file dirname $file_rev]] == -1} {
			displayMessage "CsetTool must be run in a repository"
			exit 0
		}
	}  else {
		cd2product
	}

	loadState cset
	widgets
	restoreGeometry cset

	if {$gc(cset.annotation) != ""} {
		set showAnnotations 1
	}

	if {$stdin == 1} {
		getFiles "-"
	} else {
		getFiles $revs $file_rev
	}

	bind . <Destroy> {
		if {[string match "." %W]} {
			saveState cset
		}
	}

	after idle [list wm deiconify .]
	after idle [list focus -force .]
}

main
